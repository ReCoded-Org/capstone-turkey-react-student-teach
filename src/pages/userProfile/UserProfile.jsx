import { useEffect, useState } from 'react';
import { IoSettingsOutline } from 'react-icons/io5';
import { useSelector } from 'react-redux';
import { Image } from 'cloudinary-react';
import coverPlaceholder from '../../assets/images/coverPlaceholder.jpg';
import userProfilePlaceholder from '../../assets/images/profilePlaceholer.png';
import LatestAnswers from '../../components/profile/latestAnswers/LatestAnswers';
import LatestQuestions from '../../components/profile/latestQuestions/LatestQuestions';
import ProfileSetting from '../../components/modals/profileEdit/profilePersonalInfoEdit/ProfilePersonalInfoEdit';
import ProfilePicEdit from '../../components/modals/profileEdit/profilePicEdit/ProfilePicEdit';

function UserProfile() {
  const [latestSection, setLatestSection] = useState(false);
  const [open, setOpen] = useState(false);
  const [openPicEdit, setOpenPicEdit] = useState(false);
  const [picStatus, setPicStatus] = useState([]);
  const [cloudinaryStatus, setCloudinaryStatus] = useState([]);
  const darkMode = useSelector((state) => state.darkModeReducer.darkMode);
  const avatar = useSelector(
    (state) => state.fetchAllTutorReducer.user?.avatar,
  );
  const { firstName } = useSelector((state) => state.signIn.user.userInfo);
  const { lastName } = useSelector((state) => state.signIn.user.userInfo);
  const firstNameSignUp = useSelector(
    (state) => state.signIn.signUp.isSignedUp.firstName,
  );
  const lastNameSignUp = useSelector(
    (state) => state.signIn.signUp.isSignedUp.lastName,
  );
  const formattedFirstName =
    firstName && firstName.charAt(0).toUpperCase() + firstName.slice(1);
  const formattedLastName =
    lastName && lastName.charAt(0).toUpperCase() + lastName.slice(1);
  const formattedFirstNameSignUp =
    firstNameSignUp &&
    firstNameSignUp.charAt(0).toUpperCase() + firstNameSignUp.slice(1);
  const formattedLastSignUp =
    lastNameSignUp &&
    lastNameSignUp.charAt(0).toUpperCase() + lastNameSignUp.slice(1);
  const editProfilePicStatus = useSelector(
    (state) => state.editProfilePicReducer?.message,
  );
  const uploadPicCloudinaryStatus = useSelector(
    (state) => state.uploadPicCloudinaryReducer?.data,
  );

  useEffect(() => {
    let cancel = true;
    if (cancel) {
      setPicStatus(editProfilePicStatus[editProfilePicStatus.length - 1]);
      setCloudinaryStatus(
        uploadPicCloudinaryStatus[uploadPicCloudinaryStatus.length - 1]?.status,
      );
    }

    const timer = setTimeout(() => {
      setPicStatus([]);
      setCloudinaryStatus([]);
    }, 3000);

    return () => {
      cancel = false;
      clearTimeout(timer);
    };
  }, [editProfilePicStatus, uploadPicCloudinaryStatus]);

  return (
    <div
      className={`${darkMode ? 'bg-primaryDark' : 'bg-white'}  ${
        darkMode ? 'text-white' : 'text-black'
      } lg:pb-[5rem]`}
    >
      <section className="min-h-[30vh] pt-10 lg:min-h-[50vh] flex justify-center">
        <div className="self-center">
          <div className="relative self-center">
            <img
              className="w-[50rem] h-[8rem] object-fill rounded-md lg:h-[20rem] select-none"
              src={coverPlaceholder}
              alt="cover"
            />
          </div>
          <div className="mt-3">
            <div className="relative ">
              <Image
                cloudName="eyeblinded"
                src={avatar || userProfilePlaceholder}
                className="h-[4rem] w-[4rem] object-cover ml-5 rounded-full absolute border-[2px] border-white top-[-3.3rem] lg:top-[-5rem] lg:h-[7rem] lg:w-[7rem] lg:ml-[3rem] cursor-pointer select-none"
                onClick={() => setOpenPicEdit(true)}
                alt="profile"
                aria-hidden="true"
              />
              <span
                onClick={() => setOpenPicEdit(true)}
                className="h-[6.9rem] w-[6.9rem] top-[-5rem] left-[3.1rem] bg-white absolute rounded-full opacity-0 hover:opacity-60 font-bold cursor-pointer transition-all ease-in-out duration-300 lg:flex justify-center items-center hidden text-black"
                aria-hidden
              >
                <h1 className="opacity-100">Edit</h1>
              </span>
            </div>

            <div className="flex flex-col justify-center items-center relative">
              {picStatus?.error && cloudinaryStatus === 400 ? (
                <h1 className="text-red-500 top-[-5px] absolute">
                  Failed to update picture
                </h1>
              ) : (
                ''
              )}
              {picStatus?.message && cloudinaryStatus === 200 ? (
                <h1 className="text-green-500 top-[-5px] absolute">
                  Updated successfully
                </h1>
              ) : (
                ''
              )}
              <IoSettingsOutline
                onClick={() => setOpen(true)}
                className="place-self-end mr-3 p-1 text-2xl lg:text-3xl  rounded-full text-cusOrange cursor-pointer hover:bg-cusOrange hover:text-white hover:opacity-90 ease-in-out transition-all duration-300 absolute top-[-.6rem]"
              />
              <h1 className="mt-5 font-semibold text-xl">
                {formattedFirstName || formattedFirstNameSignUp}{' '}
                {formattedLastName || formattedLastSignUp}
              </h1>
            </div>
          </div>
        </div>
      </section>
      <ProfileSetting open={open} setOpen={setOpen} />
      <ProfilePicEdit
        open={openPicEdit}
        setOpen={setOpenPicEdit}
        label="Update your picture"
      />

      <section className="text-sm md:min-h-[50vh] m-7 mb-0 lg:min-h-[70vh] lg:text-base">
        <div className="flex justify-around items-center lg:justify-center p-5 lg:mr-7">
          <button
            className={`px-3 py-2 lg:px-5 lg:py-3 lg:mr-10 ease-in-out transition-all hover:text-cusOrange ${
              !latestSection
                ? 'text-white rounded bg-cusOrange hover:text-white '
                : null
            }`}
            onClick={() => setLatestSection(false)}
            type="button"
          >
            Latest Questions
          </button>

          <button
            className={`px-3 py-2 lg:px-5 lg:py-3 ease-in-out transition-all hover:text-cusOrange lg:ml-10 ${
              latestSection
                ? 'text-white rounded bg-cusOrange hover:text-white'
                : null
            }`}
            onClick={() => setLatestSection(true)}
            type="button"
          >
            Latest Answers
          </button>
        </div>
        <div className="flex flex-col items-center justify-center pb-10 text-black">
          {!latestSection ? <LatestQuestions /> : <LatestAnswers />}
        </div>
      </section>
    </div>
  );
}

export default UserProfile;
