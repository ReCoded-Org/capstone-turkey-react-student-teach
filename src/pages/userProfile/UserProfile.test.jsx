import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import UserProfile from './UserProfile';

it('Navbar snapshot', () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <UserProfile />
      </MemoryRouter>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
