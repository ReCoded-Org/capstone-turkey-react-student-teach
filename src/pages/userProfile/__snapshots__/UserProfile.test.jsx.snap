// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar snapshot 1`] = `
<div>
  <section
    className="min-h-[30vh] mt-10 m-8 lg:min-h-[50vh] lg:mx-[0rem] lg:flex lg:justify-center"
  >
    <div>
      <div
        className="relative"
      >
        <img
          alt="cover"
          aria-hidden="true"
          className="w-[50rem] h-[8rem] object-fill rounded-md lg:h-[20rem] cursor-pointer select-none"
          onClick={[Function]}
          src="coverPlaceholder.jpg"
        />
        <svg
          className="text-white text-3xl p-1  absolute top-[1rem] right-[1rem]  hover:bg-white hover:text-cusOrange rounded-full cursor-pointer transition-all ease-in-out duration-300"
          fill="currentColor"
          height="1em"
          onClick={[Function]}
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M3 21h3.75L17.81 9.94l-3.75-3.75L3 17.25V21zm2-2.92l9.06-9.06.92.92L5.92 19H5v-.92zM18.37 3.29a.996.996 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83a.996.996 0 000-1.41l-2.34-2.34z"
          />
        </svg>
      </div>
      <div
        className="mt-3"
      >
        <div
          className="relative bg-red-400"
        >
          <img
            alt="profile"
            aria-hidden="true"
            className="h-[4rem] w-[4rem] object-cover ml-5 rounded-full absolute border-[2px] border-white top-[-3.3rem] lg:top-[-5rem] lg:h-[7rem] lg:w-[7rem] lg:ml-[3rem] cursor-pointer select-none"
            onClick={[Function]}
            src="profilePlaceholer.png"
          />
          <span
            aria-hidden={true}
            className="h-[6.9rem] w-[6.9rem] top-[-5rem] left-[3.1rem] bg-white absolute rounded-full opacity-0 hover:opacity-60 font-bold cursor-pointer transition-all ease-in-out duration-300 lg:flex justify-center items-center hidden"
            onClick={[Function]}
          >
            <h1
              className="opacity-100"
            >
              Edit
            </h1>
          </span>
        </div>
        <div
          className="flex flex-col justify-center items-center relative"
        >
          <svg
            className="place-self-end mr-3 p-1 text-2xl lg:text-3xl  rounded-full text-cusOrange cursor-pointer hover:bg-cusOrange hover:text-white hover:opacity-90 ease-in-out transition-all duration-300 absolute top-[-.6rem]"
            fill="currentColor"
            height="1em"
            onClick={[Function]}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M262.29 192.31a64 64 0 1057.4 57.4 64.13 64.13 0 00-57.4-57.4zM416.39 256a154.34 154.34 0 01-1.53 20.79l45.21 35.46a10.81 10.81 0 012.45 13.75l-42.77 74a10.81 10.81 0 01-13.14 4.59l-44.9-18.08a16.11 16.11 0 00-15.17 1.75A164.48 164.48 0 01325 400.8a15.94 15.94 0 00-8.82 12.14l-6.73 47.89a11.08 11.08 0 01-10.68 9.17h-85.54a11.11 11.11 0 01-10.69-8.87l-6.72-47.82a16.07 16.07 0 00-9-12.22 155.3 155.3 0 01-21.46-12.57 16 16 0 00-15.11-1.71l-44.89 18.07a10.81 10.81 0 01-13.14-4.58l-42.77-74a10.8 10.8 0 012.45-13.75l38.21-30a16.05 16.05 0 006-14.08c-.36-4.17-.58-8.33-.58-12.5s.21-8.27.58-12.35a16 16 0 00-6.07-13.94l-38.19-30A10.81 10.81 0 0149.48 186l42.77-74a10.81 10.81 0 0113.14-4.59l44.9 18.08a16.11 16.11 0 0015.17-1.75A164.48 164.48 0 01187 111.2a15.94 15.94 0 008.82-12.14l6.73-47.89A11.08 11.08 0 01213.23 42h85.54a11.11 11.11 0 0110.69 8.87l6.72 47.82a16.07 16.07 0 009 12.22 155.3 155.3 0 0121.46 12.57 16 16 0 0015.11 1.71l44.89-18.07a10.81 10.81 0 0113.14 4.58l42.77 74a10.8 10.8 0 01-2.45 13.75l-38.21 30a16.05 16.05 0 00-6.05 14.08c.33 4.14.55 8.3.55 12.47z"
              fill="none"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="32"
            />
          </svg>
          <h1
            className="mt-2 "
          >
            Username
          </h1>
        </div>
      </div>
    </div>
  </section>
  <section
    className="text-sm md:min-h-[50vh] m-7 lg:min-h-[70vh] lg:text-base"
  >
    <div
      className="flex justify-around items-center lg:justify-center lg:mr-7"
    >
      <button
        className="px-3 py-2 lg:px-5 lg:py-3 lg:mr-10 hover:scale-110 ease-in-out transition-all hover:text-cusOrange text-white rounded bg-cusOrange hover:text-white "
        onClick={[Function]}
        type="button"
      >
        Latest Questions
      </button>
      <button
        className="px-3 py-2 lg:px-5 lg:py-3 hover:scale-110 ease-in-out transition-all hover:text-cusOrange lg:ml-10 null"
        onClick={[Function]}
        type="button"
      >
        Latest Answers
      </button>
    </div>
    <div
      className="flex flex-col items-center justify-center lg:pb-10"
    >
      <div>
        <div
          className=" relative max-w-3xl mx-auto "
        >
          <div
            className=" mx-auto my-8 max-w-3xl px-6 pt-4 pb-6 bg-[#F0F0F0] rounded-md"
          >
            <div
              className="flex mb-5"
            >
              <img
                alt="user.png"
                className="w-10 h-10 rounded-full"
                src="https://via.placeholder.com/150"
              />
              <p
                className="pl-2 font-semibold self-center text-sm mr-auto"
              >
                hazarcann
              </p>
              <div
                className="text-xl   self-center cursor-pointer"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  onClick={[Function]}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    Object {
                      "color": undefined,
                    }
                  }
                  viewBox="0 0 192 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                  />
                </svg>
              </div>
              <div
                className="absolute -right-0 top-12 lg:-right-48 lg:top-0 z-10 w-44 rounded-lg shadow-lg  border-[#CA7560] border-[4px]  border-opacity-50 bg-white transition  py-1 lg:py-3 hidden opacity-0"
                id="dropdown"
              >
                <button
                  className=" w-full lg:w-full text-sm text-left text-gray-800 hover:text-gray-900 hover:bg-gray-200 transition px-2 lg:px-4 py-1"
                  type="button"
                >
                  Edit Question
                </button>
                <button
                  className="w-full lg:w-full text-sm text-left hover:bg-gray-200 transition px-2 lg:px-4 py-1"
                  type="button"
                >
                  Add to Bookmark
                </button>
                <button
                  className="w-full lg:w-full text-sm text-left  hover:bg-gray-200 transition px-2 lg:px-4 py-1"
                  type="button"
                >
                  Share
                </button>
                <button
                  className="w-full lg:w-full text-sm text-left text-red-600 hover:text-red-700 hover:bg-gray-200 transition px-2 lg:px-4 py-1"
                  type="button"
                >
                  Delete Question
                </button>
              </div>
            </div>
            <h3
              className="text-2xl mb-5"
            >
              <span
                className="text-[#CA7560] "
              >
                Q
              </span>
              : 
              Loeerasdsaasd
            </h3>
            <p
              className="text-sm mb-7 pl-7 line-clamp-3"
            >
              questionTexttttw
            </p>
            <div
              className="hover:scale-105 transition-transform mx-auto w-1/2 "
            />
          </div>
          <div
            className="relative w-full px-2 lg:px-20 "
          >
            <input
              className=" line-clamp-5 pr-14 md:pr-16 lg:pr-14 focus:outline-none border-[#CA7560] border-[3px] border-opacity-50 rounded-lg py-4 pl-5 text-xs w-full "
              placeholder="Answer the question..."
              type="text"
            />
          </div>
        </div>
        <div
          className=" relative max-w-3xl mx-auto "
        >
          <div
            className=" mx-auto my-8 max-w-3xl px-6 pt-4 pb-6 bg-[#F0F0F0] rounded-md"
          >
            <div
              className="flex mb-5"
            >
              <img
                alt="user.png"
                className="w-10 h-10 rounded-full"
                src="https://via.placeholder.com/150"
              />
              <p
                className="pl-2 font-semibold self-center text-sm mr-auto"
              >
                hazarcann
              </p>
              <div
                className="text-xl   self-center cursor-pointer"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  onClick={[Function]}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    Object {
                      "color": undefined,
                    }
                  }
                  viewBox="0 0 192 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                  />
                </svg>
              </div>
              <div
                className="absolute -right-0 top-12 lg:-right-48 lg:top-0 z-10 w-44 rounded-lg shadow-lg  border-[#CA7560] border-[4px]  border-opacity-50 bg-white transition  py-1 lg:py-3 hidden opacity-0"
                id="dropdown"
              >
                <button
                  className=" w-full lg:w-full text-sm text-left text-gray-800 hover:text-gray-900 hover:bg-gray-200 transition px-2 lg:px-4 py-1"
                  type="button"
                >
                  Edit Question
                </button>
                <button
                  className="w-full lg:w-full text-sm text-left hover:bg-gray-200 transition px-2 lg:px-4 py-1"
                  type="button"
                >
                  Add to Bookmark
                </button>
                <button
                  className="w-full lg:w-full text-sm text-left  hover:bg-gray-200 transition px-2 lg:px-4 py-1"
                  type="button"
                >
                  Share
                </button>
                <button
                  className="w-full lg:w-full text-sm text-left text-red-600 hover:text-red-700 hover:bg-gray-200 transition px-2 lg:px-4 py-1"
                  type="button"
                >
                  Delete Question
                </button>
              </div>
            </div>
            <h3
              className="text-2xl mb-5"
            >
              <span
                className="text-[#CA7560] "
              >
                Q
              </span>
              : 
              Loeerasdsaasd
            </h3>
            <p
              className="text-sm mb-7 pl-7 line-clamp-3"
            >
              questionTexttttw
            </p>
            <div
              className="hover:scale-105 transition-transform mx-auto w-1/2 "
            />
          </div>
          <div
            className="relative w-full px-2 lg:px-20 "
          >
            <input
              className=" line-clamp-5 pr-14 md:pr-16 lg:pr-14 focus:outline-none border-[#CA7560] border-[3px] border-opacity-50 rounded-lg py-4 pl-5 text-xs w-full "
              placeholder="Answer the question..."
              type="text"
            />
          </div>
        </div>
        <div
          className=" relative max-w-3xl mx-auto "
        >
          <div
            className=" mx-auto my-8 max-w-3xl px-6 pt-4 pb-6 bg-[#F0F0F0] rounded-md"
          >
            <div
              className="flex mb-5"
            >
              <img
                alt="user.png"
                className="w-10 h-10 rounded-full"
                src="https://via.placeholder.com/150"
              />
              <p
                className="pl-2 font-semibold self-center text-sm mr-auto"
              >
                hazarcann
              </p>
              <div
                className="text-xl   self-center cursor-pointer"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  onClick={[Function]}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    Object {
                      "color": undefined,
                    }
                  }
                  viewBox="0 0 192 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                  />
                </svg>
              </div>
              <div
                className="absolute -right-0 top-12 lg:-right-48 lg:top-0 z-10 w-44 rounded-lg shadow-lg  border-[#CA7560] border-[4px]  border-opacity-50 bg-white transition  py-1 lg:py-3 hidden opacity-0"
                id="dropdown"
              >
                <button
                  className=" w-full lg:w-full text-sm text-left text-gray-800 hover:text-gray-900 hover:bg-gray-200 transition px-2 lg:px-4 py-1"
                  type="button"
                >
                  Edit Question
                </button>
                <button
                  className="w-full lg:w-full text-sm text-left hover:bg-gray-200 transition px-2 lg:px-4 py-1"
                  type="button"
                >
                  Add to Bookmark
                </button>
                <button
                  className="w-full lg:w-full text-sm text-left  hover:bg-gray-200 transition px-2 lg:px-4 py-1"
                  type="button"
                >
                  Share
                </button>
                <button
                  className="w-full lg:w-full text-sm text-left text-red-600 hover:text-red-700 hover:bg-gray-200 transition px-2 lg:px-4 py-1"
                  type="button"
                >
                  Delete Question
                </button>
              </div>
            </div>
            <h3
              className="text-2xl mb-5"
            >
              <span
                className="text-[#CA7560] "
              >
                Q
              </span>
              : 
              Loeerasdsaasd
            </h3>
            <p
              className="text-sm mb-7 pl-7 line-clamp-3"
            >
              questionTexttttw
            </p>
            <div
              className="hover:scale-105 transition-transform mx-auto w-1/2 "
            />
          </div>
          <div
            className="relative w-full px-2 lg:px-20 "
          >
            <input
              className=" line-clamp-5 pr-14 md:pr-16 lg:pr-14 focus:outline-none border-[#CA7560] border-[3px] border-opacity-50 rounded-lg py-4 pl-5 text-xs w-full "
              placeholder="Answer the question..."
              type="text"
            />
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
